### ISSUE TITLE
Parsing without validation

### ISSUE THREAD

**Issue by kibertoad** [colinhacks/zod#1532]
Sometimes it is beneficial to be able to turn off the validation part of `safeParse` while preserving the serialization part of it.

Most common example - enabling reply validation in testing/staging, but disabling it in production for performance reasons.

If this change is approved, I can work on a PR for it.

**Comment by JacobWeisenburger** [comment]
Are you still interested in this? Or have you found a work around?

**Comment by kibertoad** [comment]
Still interested in it; sorry, missed the question.

**Comment by JacobWeisenburger** [comment]
Would you be able to explain a little more about what you are wanting? Perhaps some code examples.

Also, I'm not sure what you mean by "the serialization part of it". Could you explain what you mean?

**Comment by kibertoad** [comment]
Sure! Take a look at https://github.com/turkerdev/fastify-type-provider-zod/blob/07f5c8ff2226a5242c4fec318e6bfc312f13896e/src/index.ts#L126

This, but with possibility not to throw an error if data is invalid.

**Comment by JacobWeisenburger** [comment]
I'm not sure what that example is showing. Could you explain which part zod is not currently doing?

**Comment by kibertoad** [comment]
It is not possible to force safeParse to return parsed result regardless of whether it is valid or not.

**Comment by JacobWeisenburger** [comment]
what would be the point of returning a result regardless of whether it is valid or not?

my understanding of reason for zod is to take an unknown value and make it known. So if you are returning a value that might not have passed validation, then it is no different than just not using zod at all.

Am I not understanding something?

**Comment by kibertoad** [comment]
Performance.
Having this as a toggle allows to run strict validation in test env, and then disable it in prod for ultra-heavy lod systems without modifying the codebase to replace zod serialization with something else.

**Comment by JacobWeisenburger** [comment]
I'm not sure what you mean by "zod serialization". What are you referring to?

**Comment by kibertoad** [comment]
what does zod.safeParse return? it's not an original object that was passed for validation, from my understanding. at the very least it triggers refinements

**Comment by JacobWeisenburger** [comment]
to the best of my knowledge, that's not what [serialization](https://en.wikipedia.org/wiki/Serialization) is.

By default, `z.object().parse` returns a new object with only the keys that are specified in the schema.
So when you say "serialization", are you referring to this process?

**Comment by kibertoad** [comment]
Right. That would explain the difference between input and output, yes.

**Comment by JacobWeisenburger** [comment]
So, if I understand you correctly, you are wanting a new feature where you can disable validation, but keep the stripping?

**Comment by kibertoad** [comment]
Yes. Preservibg refinements, such as coercing, would also be nice.

**Comment by JacobWeisenburger** [comment]
[coercion](https://github.com/colinhacks/zod#coercion-for-primitives) is not a [refinement](https://github.com/colinhacks/zod#refine) as far as I know. So I'm not exactly sure what you mean.

**Comment by kibertoad** [comment]
Right. then just coercion would be great, but fully optional.
sorry, I'm not familiar with zod codebase at all, so making lots of very clueless assumptions about it. my apologies.

**Comment by JacobWeisenburger** [comment]
No problem at all. Sorry if I seem abrasive. I'm just trying to fully understand what you are asking for.

Thanks for the feature suggestion.

**Comment by JacobWeisenburger** [comment]
PRs are always welcome. But I'm not sure this will be added. Because it seems like its removing the main point of using Zod.

**Comment by CanRau** [comment]
I just came across a use case for this and am currently unsure how to best handle this without duplicating the whole schema object without any validation, probably have to do this for now to fix a bug 😅

Anyway, in my case I'm validating a form a user submits which is handled via FormData so not a simple json object as nesting is handled like this `prop.nestedProp: value` etc.

So right now if validation fails on save the whole form just gets emptied 😰 instead I'd like a way to either re-parse in the error if statement without confusing variation or a way to duplicate the existing schema dropping all validations like `string().max(500)`, so that I can return the original data as nested json.

Actually I just realized that 1 alternative in my case would be a function (which I might already have in my utils) to just convert the nested FormData to json.

Still I think this is a valid use-case and would be useful to just have.

By the way I'm using [zod-form-data (zfd)](https://github.com/airjp73/remix-validated-form/tree/main/packages/zod-form-data) to parse FormData against my zod schema, which could also be a nice built-in although zfd works fine 🤓

**Comment by JacobWeisenburger** [comment]
Please send a full reproducible code example. I'm having a hard time understanding your use case as you have described it.

**Comment by giuliano-tx** [comment]
I'm not sure this is what OP meant really. But I was thinking that we could make use of zod schemas just to transform objects. For example, sometimes you have an object that has properties you don't want to expose to the outside. If we could use Zod just to strip out those fields that would be great. I'll write an example

```typescript
const schema = z.object({ id: z.number(), title: z.string(), author: z.string() })
const bigObject = { id: 1, title: 'Title', internalId: 'secret-to-clients' }

const leanObject = schema.safeParse(bigObject);

// Here leanObject.success === false because it's missing the author, but I'd like to have a lean object like...
// { id: 1, title: 'Title' }
// So I can expose it to clients
```

**Comment by JacobWeisenburger** [comment]
Is this what you are looking for?

```ts
const schema = z.object( {
    id: z.number(),
    title: z.string(),
    author: z.string()
} )

// https://github.com/colinhacks/zod#partial
const leanSchema = schema.partial( {
    // this sets `author` to optional for `leanSchema` but doesn't change `schema`
    author: true,
} )

const bigObject = {
    id: 1,
    title: 'Title',
    internalId: 'secret-to-clients'
}

const leanObject = leanSchema.safeParse( bigObject )
leanObject.success && console.log( leanObject.data ) // { id: 1, title: 'Title' }
```

**Comment by giuliano-tx** [comment]
@JacobWeisenburger closer, but not quite there yet. In your case if `bigObject.id` is a string like `"1"` the validation would be unsuccessful and a `leanObject` will not be returned. I know you can coerce strings into numbers but I'm thinking about cases where coercing is not possible

**Comment by vodenkaj** [comment]
Is there any plan to implement this? 

The parsing option would basically try to parse the schema, and if the validation fails, it would just remove the key from the object.

Example:
```
const rawQueryParams = {
  clientId: 'someClientId',
  dealId: '123',
  leadId: { someKey: 'someValue' },
};

const queryParamsSchema = zod.object({
  clientId: zod.string(),
  dealId: zod.coerce.number().optional(),
  leadId: zod.string(),
});

const result = queryParamsSchema.safeParse(rawQueryParams);

console.log(result); // <result not parsed due to validation error>
```

This would now fail, but what If we want to only map the values according to the zod schema, but do not make any validation?

So in our use case, we would like to do something like this:
```
const rawQueryParams = {
  clientId: 'someClientId',
  dealId: '123',
  leadId: { someKey: 'someValue' },
};

const queryParamsSchema = zod.object({
  clientId: zod.string(),
  dealId: zod.coerce.number().optional(),
  leadId: zod.string(),
});

const result = queryParamsSchema.safeMap(rawQueryParams);

console.log(result); // { "clientId": 'someClientId', dealId: 123 }
```

**Comment by jesse-aesthetronics** [comment]
@vodenkaj I came across this issue and thought of something that might help. You can use [`.catch()`](https://zod.dev/api?id=catch) to provide a default value only if validation fails. Not as easy as calling a different parse function because you would need to add that to each field, but it could be helpful for you still.

```ts
const queryParamsSchema = z.object({
  clientId: z.string().catch(undefined),
  dealId: z.coerce.number().optional().catch(undefined),
  leadId: z.string().catch(undefined),
})
```

See here https://zod-playground.vercel.app?appdata=N4IgzgxgFgpgtgQxALhALwHQHsBGArGCAFwApgAdAOwAJqIAbASxkqIEkATZazMIgJ0aUA5iQCUGCAiLQSAV0ocYAMyEwOYgDRVaShPU7dMELDH4QYGSnLg4z47AAcijLJX0OpMqPMUq1Gto01PQwCByGPBh8giKe0rIKSqqU6lpUAL5iIJogAG76cjBgKADaIBTBDMyskQDkYFhwMADCTCzsHHVBumEGXNR1AIwATADM3TohYREDwNSNzQDSMACe3A1NMABqhTB11BlBGSAAurl5ZmCulCggACwYQ09DOSCMYABaWBwAskKMFDKfRgGAZIA

And in the general spirit of this issue maybe there could be something similar to `.partial()` that only applies to the output, essentially doing the `.catch(undefined)` automatically for all fields.

**Comment by colinhacks** [comment]
There's ways to do this with Zod's internal APIs, but I'm not planning to expose this as part of Zod's public API.
